### PR TITLE
LibWeb: Add PopOverInvokerElement and use it in HTMLButtonElement

### PR DESCRIPTION
**Pull request by nbvdkamp**
The `popoverTargetElement` seems to be one of the only cases of a reflected Element? attribute in the HTML spec, so I had to implement code generation for that first.

Buttons can't actually toggle popovers yet because showing/hiding popovers is not implemented yet,
but this gets us one step closer.

### CODE REVIEW

File: Meta/Lagom/Tools/CodeGenerators/LibWeb/BindingsGenerator/IDLGenerators.cpp [hunk 1]
```diff
@@ -3748,6 +3748,47 @@ JS_DEFINE_NATIVE_FUNCTION(@class_name@::@attribute.getter_callback@)
...
+                attribute_generator.append(R"~~~(
+    auto contentAttributeValue = impl->attribute(HTML::AttributeNames::@attribute.reflect_name@);
+)~~~");
+                // 3. If reflectedTarget's explicitly set attr-element is not null:
+                //      1. If reflectedTarget's explicitly set attr-element is a descendant of any of element's shadow-including ancestors, then return reflectedTarget's explicitly set attr-element.
+                //      2. Return null.
+                attribute_generator.append(R"~~~(
+    auto const explicitly_set_attr = TRY(throw_dom_exception_if_needed(vm, [&] { return impl->get_@attribute.cpp_name@(); }));
+    if (explicitly_set_attr) {
+        if (impl->shadow_including_root().is_equal_node(&explicitly_set_attr->shadow_including_root())) {
```
Comment (tcl3) [inline Meta/Lagom/Tools/CodeGenerators/LibWeb/BindingsGenerator/IDLGenerators.cpp:3772]: This should be equivalent to `&impl->shadow_including_root() == &explicitly_set_attr->shadow_including_root()`

File: Meta/Lagom/Tools/CodeGenerators/LibWeb/BindingsGenerator/IDLGenerators.cpp [hunk 2]
```diff
@@ -3748,6 +3748,47 @@ JS_DEFINE_NATIVE_FUNCTION(@class_name@::@attribute.getter_callback@)
...
+            retval = explicitly_set_attr;
+        }
+        retval = explicitly_set_attr;
+    }
+)~~~");
+                // 4. Otherwise, if contentAttributeValue is not null, return the first element candidate, in tree order, that meets the following criteria:
+                //      candidate's root is the same as element's root;
+                //      candidate's ID is contentAttributeValue; and
+                //      candidate implements T.
```
Comment (tcl3) [inline Meta/Lagom/Tools/CodeGenerators/LibWeb/BindingsGenerator/IDLGenerators.cpp:3781]: I'd add a FIXME here to make it clear that this condition isn't currently implemented

File: Meta/Lagom/Tools/CodeGenerators/LibWeb/BindingsGenerator/IDLGenerators.cpp [hunk 3]
```diff
@@ -3748,6 +3748,47 @@ JS_DEFINE_NATIVE_FUNCTION(@class_name@::@attribute.getter_callback@)
...
+                // If no such element exists, then return null.
+                // 5. Return null.
+
+                // FIXME: This works when T is Element but might need adjustment when we handle subtypes too
+                attribute_generator.append(R"~~~(
+    else if (contentAttributeValue.has_value()) {
+        retval = impl->document().get_element_by_id(contentAttributeValue.value());
```
Comment (tcl3) [inline Meta/Lagom/Tools/CodeGenerators/LibWeb/BindingsGenerator/IDLGenerators.cpp:3788]: Using `get_element_by_id()` here could return an element that doesn't share element's root.

Instead you could use `for_each_in_inclusive_subtree_of_type`:
```c++
        impl->root().for_each_in_inclusive_subtree_of_type<DOM::Element>([&](auto& candidate) {
            if (candidate.attribute(HTML::AttributeNames::id) == contentAttributeValue.value()) {
                retval = &candidate;
                return TraversalDecision::Break;
            }
            return TraversalDecision::Continue;
        });
```

File: Libraries/LibWeb/HTML/PopoverInvokerElement.idl [hunk 4]
```diff
@@ -0,0 +1,11 @@
+[MissingValueDefault=toggle, InvalidValueDefault=toggle]
+enum PopoverTargetActionAttribute {
+    "toggle",
+    "show",
+    "hide"
+};
+
+interface mixin PopoverInvokerElement {
```
Comment (tcl3) [inline Libraries/LibWeb/HTML/PopoverInvokerElement.idl:8]: Could we add a spec link here: https://html.spec.whatwg.org/multipage/popover.html#popoverinvokerelement

File: Libraries/LibWeb/HTML/PopoverInvokerElement.idl [hunk 5]
```diff
@@ -0,0 +1,11 @@
+[MissingValueDefault=toggle, InvalidValueDefault=toggle]
```
Comment (tcl3) [inline Libraries/LibWeb/HTML/PopoverInvokerElement.idl:1]: Could we add a spec link here: https://html.spec.whatwg.org/multipage/popover.html#attr-popovertargetaction

File: Libraries/LibWeb/HTML/PopoverInvokerElement.h [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+class PopoverInvokerElement {
+public:
+    PopoverInvokerElement() { }
+
+    GC::Ptr<DOM::Element> get_popover_target_element() { return m_popover_target_element; }
+
+    void set_popover_target_element(GC::Ptr<DOM::Element> value) { m_popover_target_element = value; }
+
+private:
+    GC::Ptr<DOM::Element> m_popover_target_element;
```
Comment (tcl3) [inline Libraries/LibWeb/HTML/PopoverInvokerElement.h:23]: This needs to be visited by a `visit_edges()` method.

**Review by tcl3 — COMMENTED** [review]
I have a couple of suggestions for the codegen. I didn't have time to look at the second commit in detail.

File: Libraries/LibWeb/HTML/PopoverInvokerElement.h (hunk 6 above)
Comment (nbvdkamp) [reply Libraries/LibWeb/HTML/PopoverInvokerElement.h:23]: When trying to apply the GC visit_edges pattern I'm seeing elsewhere by making PopoverInvokerElement inherit from `JS:Cell` I run into the problem that `HTMLButtonElement` is then inheriting `Cell` from two different base classes (PopoverInvokerElement  and Object). This gives the following errors:
```
LibJS/Runtime/Realm.h:39:9: error: ‘JS::Cell’ is an ambiguous base of ‘Web::HTML::HTMLButtonElement’
   39 |         static_cast<Cell*>(object)->initialize(*this);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~
```

```
LibGC/Heap.h:47:17: error: ‘GC::Cell’ is an ambiguous base of ‘Web::HTML::HTMLButtonElement’
   47 |         return *static_cast<T*>(memory);
      |                 ^~~~~~~~~~~~~~~~~~~~~~~
```
Is there an established way of getting around this?

**Comment by nbvdkamp** [comment]
Thanks for the review, I've pushed the changes

**Comment by lukewarlow** [comment]
> The popoverTargetElement seems to be one of the only cases of a reflected Element? attribute in the HTML spec, so I had to implement code generation for that first.

There's some aria properties that need to reflect elements too so this work will be useful in future.

**Review by tcl3 — APPROVED** [review]

**Pull request merged**